src/Microsoft.Extensions.Configuration.AzureAppConfiguration/ConfigurationClientManager.cs

```diff
@@ -57,19 +109,87 @@ internal ConfigurationClientManager(IList<ConfigurationClientWrapper> clients)
-        }
+            DateTimeOffset now = DateTimeOffset.UtcNow;
+
+            // Principle of refreshing fallback clients:
+            // 1.Perform initial refresh attempt to query available clients on app start-up.
+            // 2.Refreshing when cached fallback clients have expired, FallbackClientRefreshExpireInterval is due.
+            // 3.At least wait MinimalClientRefreshInterval between two attempt to ensure not perform refreshing too often.
 
+            if (_replicaDiscoveryEnabled &&
+                now >= _lastFallbackClientRefreshAttempt + MinimalClientRefreshInterval && 
+                (_dynamicClients == null ||
+                now >= _lastFallbackClientRefresh + FallbackClientRefreshExpireInterval))
+            {
+                _lastFallbackClientRefreshAttempt = now;
+
+                await RefreshFallbackClients(cancellationToken).ConfigureAwait(false);
+
+                _lastFallbackClientRefresh = now;
+            }
+
+            foreach (ConfigurationClientWrapper clientWrapper in _clients)
+            {
+                if (clientWrapper.BackoffEndTime <= now)
+                {
+                    yield return clientWrapper.Client;
+                }
+            }
+
+            // No need to continue if replica discovery is not enabled.
+            if (!_replicaDiscoveryEnabled)
+            {
+                yield break;
+            }
+
+            foreach (ConfigurationClientWrapper clientWrapper in _dynamicClients)
+            {
+                if (clientWrapper.BackoffEndTime <= now)
+                {
+                    yield return clientWrapper.Client;
+                }
+            }
+
+            // All static and dynamic clients exhausted, check if previously unknown
+            // dynamic client available
+            if (now >= _lastFallbackClientRefreshAttempt + MinimalClientRefreshInterval)
+            {
+                _lastFallbackClientRefreshAttempt = now;
+
+                await RefreshFallbackClients(cancellationToken).ConfigureAwait(false);
+
+                _lastFallbackClientRefresh = now;
+            }
+
+            foreach (ConfigurationClientWrapper clientWrapper in _dynamicClients)
+            {
+                if (clientWrapper.BackoffEndTime <= now)
+                {
+                    yield return clientWrapper.Client;
+                }
+            }
+        }
+
         public void UpdateClientStatus(ConfigurationClient client, bool successful)
         {
             if (client == null)
             {
                 throw new ArgumentNullException(nameof(client));
             }
 
-            ConfigurationClientWrapper clientWrapper = _clients.First(c => c.Client.Equals(client));
+            ConfigurationClientWrapper clientWrapper = _clients.FirstOrDefault(c => c.Client.Equals(client));
+
+            if (_dynamicClients != null && clientWrapper == null)
+            {

```

Comment:
Kind of odd that we use `Equals` here, but `==` in GetEndpoint method.

Reply:
All use `==` is fine.